### PR TITLE
feat(infra): migrate to Next.js infrastructure

### PR DESCRIPTION
**Pull request by ovflowd**
This PR is a draft to migrate the underlying infrastructure from Handlebars and Metalsmith to JSX (React) and Next.js.

## PR-Progress (Checklist)

* [x] Migrating the Styles and Layouts
    * [x] Migrate the Handlebars template engine scripts into JSX Hooks/JavaScript utilities
        * Around **100%** of the scripts are now migrated to Next.js. The remaining scripts will be ported together with their respective Layouts.
    * [x] Migrating the Handlebars layouts (templates) into JSX/React Components
      * Around **100%** of the layouts are migrated and ready.
    * [x] Re-linking the existing styles (Sass) to the new JSX Components
* [x] Update the Nextra Theme with:
    * [x] Understanding the current Frontmatter
    * [x] Linking the correct layouts based on Frontmatter
    * [x] Update the `theme.tsx` to use the correct layout
* [x] Migration of Content
    * [x] All content is buildable
    * [x] Non-translated pages within translation folders were removed (As they serve no use)
    * [x] Markdown build errors were fixed and non-conforming Markdown was fixed
* [x] Final underlying mechanisms such as:
    * [x] Migrating the pure JavaScript scripts (public/static/scripts) to be part of JSX
      * I've decided to do this outside this PR, as it would be changing the content, and this can be done as they will be discarded in the future anyways.
    * [x] Remove old/legacy scripts/
    * [x] Update other scripts, such as the release-post script
    * [x] Add the meta/header/metadata tags with Nextra
    * [x] Implement the RSS feed feature
    * [x] Implement the Sitemap feature
    * [x] i18n functionality
      * [x] Port all translations from their JSON files to the`i18n/locales/{localeCode}.json`
    * [x] Node.js version & release schedule metadata
* [x] Preparing the PR to its readiness for review
    * [x] Check the Markdown files are OK and working with the current MDX engine
    * [x] Update the tests and final preparations
    * [x] Check if everything is working, such as build, links, and other things
    * [x] Re-enable build on CI
    * [x] Optimise legacy scripts from `public/static/js` to be modern-proof and work well with Next.js
    * [x] Moved the `external/survey` content to `public/en/`
* [x] PR ready to be reviewed
* [ ] Performance Optimisations **(Updated on 24/01/2023)**
    * [ ] Compression and modification of MDX HTML output on Next.js-side
    * [x] Reduction of build-time and bundled content
    * [x] Compression of the bundled i18n JSON

### Migration of Structure

A few directories have been changed to accommodate the new Next.js folder structure. Notably, these are the significant changes:

- Some Handlebar utilities became React Components under `components`
- Sass files (SCSS) were moved to `styles`
- Pages were moved from `locale/` to `pages/`
- Static Files were moved from `static/` to `public/static` (Links will be preserved)
- Layouts will be ported over React to `layouts`
- Data files (such as versions.json and schedule.json) were moved to `data`
- I18n Configuration files and Translations will be moved to `i18n`

### The deletion of per-locale `site.json` and creation of a global `site.json`

i18n specific metatdata will be moved to `i18n/config.json` as the aggregate of all i18n metadata. Specific translation texts (such as "Next" button) or other translated text will be moved to `i18n` localization files such as `i18n/locales/en.json`. This is the same approach adopted currently on `nodejs.dev` repository to ease the migration.

Navigation structure and other non-related global metadata will be moved towards a global `data/site.json` and respectively for Navigation to `pages/_meta.json` files.

### Migration of Handlebars Templates to JSX

Many of the current "Handlebars" scripts (for templating-engine) will be ported to either JavaScript utils or JSX Components, and the current Layouts will also be ported to JSX.

### Migration from old tooling/scripts

Currently, we have scripts to generate the metadata of releases, versions, and the release schedule of Node.js; we should adopt the nodes from Nodejs.dev as they're going to be adopted anyways. This can be done either in this PR or in a follow-up, depending on the complexity of migrating the current scripts.

## Items to be done after this PR

As mentioned, this is an initial implementation for [this issue](https://github.com/nodejs/nodejs.org/issues/4958), following-up there will still be some improvements to be done, such as removing more legacy content and normalizing the JSX code and a few other things.

After that, we can start porting the layout and content from the [Nodejs.dev repository](https://github.com/nodejs/nodejs.dev).

## Final Considerations **(Updated on 24/01/2023)**

As mentioned, this is an initial implementation (part of a transitionary phase). Much of the content added within this PR will be discarded in future PRs.

A few quirks from the original nodejs.org website are not available within this initial PR, as they do not bring value in the short-term (having the consideration that these same things will be discarded and considering the value x effort to have them done).

### What was not ported from the original Nodejs.org codebase?

- Edit on GitHub links. It used a complex logic that would need to be replicated differently, which would be too complex and non-ideal for this current implementation.
  - Edit Links will be available in future PRs from the ported codebase of Nodejs.dev
- Non-translated pages will not be fallback to the English page + translated React Components
  - On the `main` branch, all non-translated pages have their respective i18n context injected and the original English content file (Markdown) copied and bundled together.
  - This would allow you to access `nodejs.org/{otherLanguage}/some-page-not-translated-in-that-language` without being redirected to `https://nodejs.org/en/the-original-page-in-english`
  - This was an exciting gimmick as at least the menubars, and other static translatable content would be translated.
  - Yet this is not ideal as we would have hundreds of new pages that, in the end, do not bring much context and value to that language for just the benefit of having a few buttons and header texts translated. With the enormous cost of performance and amount of generated files.
  - **The routing of non-translated pages is handled by Nginx, meaning if the page does not exist on a translated version, it will be fallback to English.**
  - **This will be addressed on future PRs with SSR (Server-Side-Rendering) and the usage of Next.js's built-in i18n feature.**
- `Download Page => Previous Builds`, instead of showing all versions since `Node.js 0.10`, we adopted the same approach done on `https://nodejs.dev` and show all the previous versions that are still "maintained" or in "maintenance" mode. All that page data was significant, and we wonder if people want to search for such old versions of Node.js on the website.
- The Handlebars Unit Tests were removed. As these layouts and components are soon to be removed, we figured that porting a few unit tests would not bring benefits.
- `HTML` tag with `dir` and `htmlLang` attributes cannot be added due of how `Next.js` document composition works. This will be solved with SSR on a following PR.

### What is going to be done on very immediate PRs

- All the current content (except `en/blog`) will be replaced by the pages from `nodejs.dev`. Meaning translations will also be dropped. This was [discussed here](https://github.com/nodejs/nodejs.org/issues/5001).
- We're planning on adopting either Vercel infrastructure (which would remove a lot of burden on our Build WG) or just a regular SSR installation on our servers. This would allow ISB/ISR (Incremental Static Builds/Rendering) and SSR (Server Side Rendering). It would also allow a better i18n implementation that doesn't require the `two-letter` folder approach and way less file generation at build time.
- Once the migration of the layouts from `nodejs.dev` starts, all the current layouts, scripts, and many other existing components will be dropped or adapted.
- OS detection, Dark Mode toggle, and Language Picker are done through legacy vanilla JS scripts that are not-minified/optimized. These will be removed in favor of JSX code that comes from `nodejs.dev`

### Testing Locally / Caveats on the Static Build

**TL;DR:** `npm ci && npm run serve`. **Browser:** `http://localhost:3000/en/`

- A few things are handled directly by NGINX on our Infrastructure. Such as `/` to `/en` redirection, 404-page handling, and redirection from untranslated pages to `/en/original-apge`. So please do not worry if these don't work if you're trying out a complete exported build of this branch locally. These should work out of the box with the current NGINX configuration.
- To run the current branch on development mode, we recommend `npm ci && npm run serve`
- To run the current branch on production mode, we recommend `npm ci && npm run build && npm run start`

---

**Live version:** (Updated on 23/01/2023) https://nodejs-org-next-92b1511700.netlify.app/en/ 

Relates to https://github.com/nodejs/nodejs.org/issues/4958

cc @nodejs/website @nodejs/nodejs-dev @nodejs/build

### CODE REVIEW

**Comment by richardlau** [comment]
> This PR is a draft to migrate the underlying infrastructure from Handlebars and Metalsmith to JSX (React) and Next.js.
> ## Changes
> ### Migration of Structure
> 
> A few directories are changed to accommodate the new Next.js folder structure. Notably, these are the significant changes:
> 
>     * Pages were moved from `locale/` to `pages/`
> ...
> Since this PR is in a draft state, there are still a few things to be done before we can say we're ready to get this PR reviewed:
> 
> ...
>     * Move all blog posts to the new `pages` folder and new file format and update its Frontmatter
> 

Can you add updating [`scripts/release-post.js`](https://github.com/nodejs/nodejs.org/blob/main/scripts/release-post.js) to the list of things to be done. From a @nodejs/releasers POV, we want running that script (to generate the blog posts for releases) to create the post in the correct folder.

**Comment by ovflowd** [comment]
Oh @richardlau I'm going to inspect the whole scripts folder to remove any non-needed script and update all the needed ones. Don't worry about that 👀

But yea, going to update the PR description.

**Comment by tniessen** [comment]
Do we have any insights into how this will affect performance, file sizes, latency, etc.?

Do URLs to existing resources remain the same?

**Comment by ovflowd** [comment]
@tniessen this is only an infrastructure migration meaning:

- layouts and content should be the same
- all existing URLs should be the same
- deployment should also be the same without anything being changed on any config in the server side.

This means this PR is a foundational change for:

- A new framework behind the website
- Foundational milestone for migrating the layout, content and features from nodejs.dev over here
- Allowing all blazing-edge features from Next.js
- The way how the website is built, bundled and minified is different.

In terms of DX and development this means:

- Many things such as SEO, loading of components, linting, the set of features, and many other things are managed by Next.js
- Instead of pure HTML templates with "php" era of templating engine are now replaced by JSX/React
- Markdown is now standardised and supports MDX
- Easier to develop new features and styling as support of CSS modules, React, and et cetera

For content publishers:
- Creating Markdown pages should be the same experience
- Translating pages is the same experience 
- Translating component translation strings now is done on a ICU-alike format

**Comment by ovflowd** [comment]
For performance and how long build takes time, this is something we can only benchmark once the whole thing is buildable. Right now it doesn't build yet, so there are a lot of things to be done :)

But of course I expect instead of taking less than one minute to at least 2-7 minutes for full cold-build time. (First time it gets built)

For CI we will be leveraging cache and other things but it's honestly too soon to talk about any of these.

**Comment by nschonni** [comment]
You can drop me from the website team after this lands

**Comment by ovflowd** [comment]
@nschonni ... why? And any reason why this is being commented in this PR? 😅 (feels off-topic or targeted comment)

**Comment by nschonni** [comment]
This will remove the last of the parts where I contributed much here, so it's just a task for after this lands to remove me from this team.

**Comment by ovflowd** [comment]
I see, honestly speaking you're always welcome to contribute 🙃

I don't think you _need_ to be removed from the team. You've always given valuable insights and comments on the nodejs.dev repo, so even if some (or the last pieces) of your code are being removed, your contributions are definitely still valued and future contributions even if they are only with PR reviews are always valued and welcome!

But of course, if you want to leave, that's up to you. 🥲

<sup>sorry for the off-topic 🫶</sup>

**Comment by nschonni** [comment]
Thanks, I'm still on several other teams in the org, but I don't have interest in MDX/React, so I don't need the extra notifications going forward

**Comment by ovflowd** [comment]
As this PR it's getting closer and closer to a ready state, here's an update of what it's left to be done:

* Migrating the Styles and Layouts
    * Migrate the Handlebars template engine scripts into JSX Hooks/JavaScript utilities
    * Migrating the Handlebars layouts (templates) into JSX/React Components
    * Re-linking the existing styles (Sass) to the new JSX Components
* Update the Nextra Theme with:
    * Understanding the current Frontmatter
    * Linking the correct layouts based on Frontmatter
    * Update the `theme.tsx` to use the right layout
* Final underlying mechanisms such as:
    * Migrating the pure JavaScript scripts (public/static/scripts) to be part of JSX
    * Remove old/legacy scripts/
    * Update other scripts such as the release-post script
    * Add the meta/header/metadata tags with Nextra
* Preparing the PR to its readiness for review
    * Check the Markdown files are OK and working with the current MDX engine
    * Update the tests and final preparations
    * Check if everything is working, such as build, links, and other things
    * Re-enable build on CI
* PR ready to be reviewed

**Comment by ovflowd** [comment]
**Sticky Notes:** (So I don't forget):

- https://github.com/jpmonette/feed (For the RSS implementation)
- https://www.npmjs.com/package/next-sitemap (For the Sitemap Implementation)

File: .remarkrc
Comment (nschonni) [inline .remarkrc]: This is the style used in the org and API docs, that's why it's set in the `remark-preset-lint-node`

Comment (ovflowd) [reply .remarkrc]: Gotcha. Going to undo this. It was generating a few warnings for the recently touched files, and I was wondering why it was not complaining before 🤔

Comment (nschonni) [reply .remarkrc]: Most of those rules should eventually go on, but there are a few that are incompatible. It's been awhile, and I don't recall if that JSON-ish config file actually supported comments, so it's unfortunatly not obvious

Comment (nschonni) [reply .remarkrc]: Prettier using remark to do it's formatting, but unfortunately doesn't read from the config file, so you may be getting conflicts between the 2

File: crowdin.yml
Comment (AugustinMauroy) [inline crowdin.yml]: Knowledge base was remove you can remove it

File: .remarkrc
Comment (nschonni) [reply .remarkrc]: BTW, adding the prettier preset is also just overriding the org style

Comment (ovflowd) [reply .remarkrc]: > BTW, adding the prettier preset is also just overriding the org style

I checked the docs, and the prettier preset is actually used to remove the conflicts that prettier has by default. (Like Prettier doesn't understand what Frontmatter is... It should not conflict with our rules afaik)

Comment (nschonni) [reply .remarkrc]: It is disabling rules/config that are explicitly turned on by the Node preset. They may conflict with prettier, but that is Node markdown styleguide (today). If you want to be more prettier compatible in the org, you can suggest changes in https://github.com/nodejs/remark-preset-lint-node, but those changes also need to be make to the API docs in nodejs/node

Comment (ovflowd) [reply .remarkrc]: Are you sure we have conflicting rules? Where do you see the conflicts? 👀

Comment (nschonni) [reply .remarkrc]: The unorderd list style is the first one I saw, but you'd need to do a crosswalk of https://github.com/nodejs/remark-preset-lint-node/blob/main/index.js (including the remark-preset-lint-recommended list) to see the full conflict.
Easiest way might be to move the prettier plug-in load before node one too see the Node styles re-enabling all the rules it has disabled, but in the end, not using the plug-in is the answer

Comment (ovflowd) [reply .remarkrc]: Question: If I set the remark-lint-node rule below the prettier one, the remark-lint-node will override prettier rules, right?

Comment (ovflowd) [reply .remarkrc]: Ah your answer just answered my question 🙈 let me try that then ;D

Comment (ovflowd) [reply .remarkrc]: @nschonni sadly, for now, we need to keep things as they are. Otherwise, linting gets wholly borked. We can check with the TSC regarding updating our listing rules for Markdown files.

**Comment by ovflowd** [comment]
Happy to announce this PR is finally ready to be reviewed 🎉

**Comment by AugustinMauroy** [comment]
Is it me or is the toggle dark mode not working?

**Comment by bnb** [comment]
@ovflowd is there a built demo version?

**Comment by ovflowd** [comment]
> @ovflowd is there a built demo version?

Sadly there's not, I believe I could set-up a GitHub pages, but it would not "route" the `.html` files I guess? Any advise here?

**Comment by ovflowd** [comment]
> Is it me or is the toggle dark mode not working?

@AugustinMauroy it seems to be working fine for me.

File: README.md
Comment (shanpriyan) [inline README.md]: I think this should point to the username of the person who forked the repo instead of nodejs itself

**Comment by mikeesto** [comment]
Props for the effort on this. FWIW, the theme and language pickers don't work for me on the news (/en/blog) page but work fine elsewhere. I also visually compared most of the other pages to the current live version and couldn't see any obvious issues. `/de/blog/` and other locales show English for the blog/news page currently, but on this PR it's a 404 - not sure if that's intended

**Comment by targos** [comment]
~How can I test it locally? both `npm run build` and `npm run serve` fail with ``Error: > Couldn't find a `pages` directory. Please create one under the project root``~

Edit: sorry, I checked out the wrong branch (`feat/next-js`)

**Comment by targos** [comment]
Here's a build: https://nodejs-org-next-92b1511700.netlify.app/en/

**Comment by ovflowd** [comment]
> Props for the effort on this. FWIW, the theme and language pickers don't work for me on the news (/en/blog) page but work fine elsewhere. I also visually compared most of the other pages to the current live version and couldn't see any obvious issues. /de/blog/ and other locales show English for the blog/news page currently, but on this PR it's a 404 - not sure if that's intended

Hey, @mikeesto I've just updated the main PR description explaining about these behaviors. But a TL;DR is: As this is a fully static generated website, the i18n redirection (pages that not exist within a certain language) is done via NGINX on our production infrastructure.

**Comment by ovflowd** [comment]
> Here's a build: https://nodejs-org-next-92b1511700.netlify.app/en/

@targos any way we could add the current build wg nginx config within this netlify app? (I doubt)

**Comment by ovflowd** [comment]
**Hey y'all, I've just updated the main description of this PR containing more details regarding divergences within this implementation, immediate next steps, etc**

**Comment by AliTechGeek52** [comment]
@ovflowd Can I contribute to this as well with you ?

**Comment by ovflowd** [comment]
> @ovflowd Can I contribute to this as well with you ?

What do you mean?

**Comment by targos** [comment]
@ovflowd 

> any way we could add the current build wg nginx config within this netlify app? (I doubt)

probably not. I just used [Netlify Drop](https://app.netlify.com/drop) to upload the build folder.

**Comment by ovflowd** [comment]
> probably not. I just used [Netlify Drop](https://app.netlify.com/drop) to upload the build folder.

Sounds alright to me, I added the caveats of what we should expect when testing without nginx. (On the PR description)

**Comment by AliTechGeek52** [comment]
> > @ovflowd Can I contribute to this as well with you ?
> 
> What do you mean?

My mean is can I also contribute on this pr ?

**Comment by ovflowd** [comment]
> My mean is can I also contribute on this pr ?

There's nothing to be done left in this PR, and sorry, this PR is somewhat highly-sensitive, so I would rather keep the work for known contributors. But you're of course free to review and comment and give your opinions on this PR :)

**Comment by ovflowd** [comment]
_that moment you accidentally close the PR_

File: README.md
Comment (shanpriyan) [inline README.md]:
```suggestion
  - `npm run prettier:fix` attempts to fix any style errors
```

**Comment by bmuenzenmeyer** [comment]
Fired it up and will be checking on Mac and Windows today. So far so good!

A small suggestion to add to the bottom PR body - it might be helpful to reviewers to tell them that the should visit http://localhost:3000/en/ directly. I know you mention the nginx rerouters not working, but I had to run the command and then read through the PR body carefully to notice that.

File: COLLABORATOR_GUIDE.md
Comment (bmuenzenmeyer) [inline COLLABORATOR_GUIDE.md]: This content was deleted but the headings remain here
```suggestion
```

**Review by bmuenzenmeyer — CHANGES_REQUESTED** [review]

File: README.md
Comment (ovflowd) [reply README.md]:
```suggestion
git clone https://github.com/{your username}/nodejs.org
```

**Comment by tniessen** [comment]
> > Is it me or is the toggle dark mode not working?
> 
> @AugustinMauroy it seems to be working fine for me.

In the preview build, dark mode keeps resetting when I switch between pages in Firefox.

---

No strong opinion on this PR in general.

**Comment by ovflowd** [comment]
> In the preview build, dark mode keeps resetting when I switch between pages in Firefox.

Interesting, I'm not aware of any code regarding storing this anywhere, but I can double check.

**Comment by AugustinMauroy** [comment]
> In the preview build, dark mode keeps resetting when I switch between pages in Firefox.

in safari the problem is not present.

**Comment by ovflowd** [comment]
> Interesting, I'm not aware of any code regarding storing this anywhere, but I can double check.

I've just fixed the issue, indeed we were not reading the localStorage.

> in safari the problem is not present.

I was able to reproduce it on Safari too. Issue was actually real :)

**Comment by AugustinMauroy** [comment]
@ovflowd I can't reproduce this issue on safari. 

>Safari Version 16.2 (18614.3.7.1.5)

https://user-images.githubusercontent.com/97875033/214385116-44c2e881-0e5a-47ed-a6d2-56a6e3be1ff7.mov

**Comment by bmuenzenmeyer** [comment]
Trying to investigate more but I haven't been able to get this building yet on Windows

```
/c/src/nodejsorg (feat/nextjs-infrastructure) 🔥 npm run build

> nodejs.org@1.0.0 build
> next build

(node:25608) ExperimentalWarning: Importing JSON modules is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
info  - Linting and checking validity of types  
info  - Creating an optimized production build  
Failed to compile.

C:src
odejsorg        heme.tsx
Module build failed: UnhandledSchemeError: Reading from "C:src
odejsorg        heme.tsx" is not handled by plugins (Unhandled scheme).
Webpack supports "data:" and "file:" URIs by default.
You may need an additional plugin to handle "c:" URIs.
    at C:\src\nodejsorg\node_modules\next\dist\compiled\webpack\bundle5.js:28:395974
    at Hook.eval [as callAsync] (eval at create (C:\src\nodejsorg\node_modules\next\dist\compiled\webpack\bundle5.js:13:28771), <anonymous>:6:1)
    at Hook.CALL_ASYNC_DELEGATE [as _callAsync] (C:\src\nodejsorg\node_modules\next\dist\compiled\webpack\bundle5.js:13:25925)
    at Object.processResource (C:\src\nodejsorg\node_modules\next\dist\compiled\webpack\bundle5.js:28:395899)       
    at processResource (C:\src\nodejsorg\node_modules\next\dist\compiled\webpack\bundle5.js:1:268020)
    at iteratePitchingLoaders (C:\src\nodejsorg\node_modules\next\dist\compiled\webpack\bundle5.js:1:267379)        
    at runLoaders (C:\src\nodejsorg\node_modules\next\dist\compiled\webpack\bundle5.js:1:271283)
    at NormalModule._doBuild (C:\src\nodejsorg\node_modules\next\dist\compiled\webpack\bundle5.js:28:395761)        
    at NormalModule.build (C:\src\nodejsorg\node_modules\next\dist\compiled\webpack\bundle5.js:28:397789)
    at C:\src\nodejsorg\node_modules\next\dist\compiled\webpack\bundle5.js:28:81243

Import trace for requested module:
C:src
odejsorg        heme.tsx
```

I'd be happy to discover it's environment-related. Are other Windows users successful yet? Oversharing in case it's useful.

**Comment by ovflowd** [comment]
@bmuenzenmeyer as you can see, even our Windows-CI is failing. No idea why. @shuding is investigating.

File: next.config.mjs
Comment (mikeesto) [inline next.config.mjs]:
```suggestion
  theme: 'theme.tsx',
```
I think this will fix the windows CI issue (at least it builds now on my fork with this change)

Comment (ovflowd) [reply next.config.mjs]: Thanks for the fix, @mikeesto! That was actually really simple!

**Comment by bmuenzenmeyer** [comment]
https://github.com/nodejs/nodejs.org/pull/4991/commits/55a96453dbfc4041da6d30738128306b9276d1b2 resolved my issue on windows too ✅

File: next.config.mjs
Comment (shuding) [reply next.config.mjs]: Oh wow, makes a lot of sense! Thank you!

File: .nvmrc
Comment (bmuenzenmeyer) [inline .nvmrc:1]: nit: this is a mutable target, similar to using `"react": "^18.9.0",` in a `package.json` but WITHOUT the benefit of a lockfile. On my teams we pin this to an explicit version.

Tools like nvm that read this and automatically switch versions will re-download Node.js for users upon traversing to this directory. A good thing, perhaps, but also noisy.

If this feels like too much of a burden to keep updating, disregard.

File: scripts/next-data/getBlogData.mjs
Comment (bmuenzenmeyer) [inline scripts/next-data/getBlogData.mjs]:
```suggestion
  // and then prepend the category name to it
```

**Review by bmuenzenmeyer — APPROVED** [review]
Noticed a metalsmith reference within `scripts/release-post/index.mjs`
Might be worth checking out since that script was mentioned in the PR as touched.

Two small things in the review, otherwise I am ✅ and suspect anything else can be tackled with fast-follower PRs.

File: .nvmrc
Comment (ovflowd) [reply .nvmrc:1]: TL;DR, we want to keep in sync that the prod version of Node is the same as the local one. IMHO, it makes error reporting and debugging somewhat easier by discarding Node.js versions of the equation.

**Comment by ovflowd** [comment]
> Noticed a metalsmith reference within scripts/release-post/index.mjs
Might be worth checking out since that script was mentioned in the PR as touched.

It's just a reference on a string. (Nothing harmful). The script is working as intended (it was heavily tested), I appreciate the note, tho.

**Comment by ovflowd** [comment]
> Two small things in the review, otherwise I am ✅ and suspect anything else can be tackled with fast-follower PRs.

Would any reviewer dive deep into the JSX code changes, Next.js configuration, and the decided approaches for asynchronous data fetching or stuff? I'm open to having reviews in that direction. But most of the reviewers might focus on comparing end-to-end behavior to validate that this transition works as intended. (Not to mention that we need the reviewer's ACK for some of the technical decisions done here).

For worse or better, code-wise, I'm pretty confident that we're following best React practices (besides when we explicitly mentioned with `Note.:` or `TODO` comments that these are temporary solutions). 

Same for the Next.js aspect, as we have people from the Next.js team working on this PR very closely. (At least I'd consider myself someone that holds an expertisè when it comes to ECMAScript and React, but of course, there are many others with way vaster knowledge than mine)

**Comment by ovflowd** [comment]
Just a FYI @richardlau @MaledongGit I expect reviews (even if they are rubber stamps) from both of you, and at least one from someone of @nodejs/build and @nodejs/tsc

**Review by AugustinMauroy — APPROVED** [review]
Nice work Claudio and other people! 🎉🥳
I can't find any other bug/issue.
> The tests were done on MacOS 13.1; Safari Version 16.2 (18614.3.7.1.5); node v18.9.1

File: package.json
Comment (bmuenzenmeyer) [inline package.json:62]: this dependency is dead - no longer used in any scripts

Comment (ovflowd) [reply package.json:62]: True, we could remove `cross-env`. Critter is also something we could remove for now, but will be used in the future for Service Workers (SW) with Google's Workbox, to enable offline-mode, PWAs, and other things.

**Comment by nschonni** [comment]
Few random thoughts, feel free to ignore:
- Might be work creating a snapshot of the old website and host it at `old.nodejs.org` with a warning banner. Could probably restore some of the other content like the deleted Knowlegebase articles that had a few issues opened about
- I think the Learn stuff done over on .dev is nice, but I think keeping that site/repo around for that content to live separately might make sense.
- Before pulling the docs content into this repo, it might be worth looking at leveraging https://github.com/nodejs/i18n as a spot to create a new `docs.nodejs.org`
- On the Blog site, it appears to be pulling down duplicate lists in the JSON data a few times like https://nodejs-org-next-92b1511700.netlify.app/_next/data/DwuP8AiWZb_pTzCXU4_eh/en/blog/release/v19.4.0.json

Overall, look and feel seems the same

**Comment by AugustinMauroy** [comment]
> Might be work creating a snapshot of the old website and host it at old.nodejs.org with a warning banner. Could probably restore some of the other content like the deleted Knowlegebase articles that had a few issues opened about

And what do you thinks about create some new content. I mean we can make the same idea of knwoledege base but more modern (I thinks opening an issue about it's a good idea).

> Before pulling the docs content into this repo, it might be worth looking at leveraging https://github.com/nodejs/i18n as a spot to create a new docs.nodejs.org

Why not create an repo like "nodejs-docs" ? (an issue should be an good idea)

**Comment by ovflowd** [comment]
> On the Blog site, it appears to be pulling down duplicate lists in the JSON data a few times like [nodejs-org-next-92b1511700.netlify.app/_next/data/DwuP8AiWZb_pTzCXU4_eh/en/blog/release/v19.4.0.json](https://nodejs-org-next-92b1511700.netlify.app/_next/data/DwuP8AiWZb_pTzCXU4_eh/en/blog/release/v19.4.0.json)

@nschonni, you're aware this is an old build, right? This was already addressed.

> Might be work creating a snapshot of the old website and host it at old.nodejs.org with a warning banner. Could probably restore some of the other content like the deleted Knowlegebase articles that had a few issues opened about

Not related to this PR.

> I think the Learn stuff done over on .dev is nice, but I think keeping that site/repo around for that content to live separately might make sense.

This is off-topic to this PR. Also, not sure you're on the loop, but the goal is to bring all the content of nodejs.dev over here. Including layout and et cetera.

> Before pulling the docs content into this repo, it might be worth looking at leveraging [nodejs/i18n](https://github.com/nodejs/i18n) as a spot to create a new docs.nodejs.org

Also off-topic.

@nschonni @AugustinMauroy I'm marking your comments as off-topic as this, honestly, is not the place to discuss what you all are talking about. For all those things mentioned, please open a discussion thread. Many of these items were already discussed in other issues, discussion threads, or channels before, so I'd appreciate it if all of you could investigate before opening duplicated content. Thank you!

---

I forgot to mention, but of course, I appreciate your thoughts, @nschonni, just probably not the right place (here this PR) as we want to keep this discussion as clean and on-topic as possible 🙇

**Comment by mcollina** [comment]
> We're planning on adopting either Vercel infrastructure (which would remove a lot of burden on our Build WG) or just a regular SSR installation on our servers. This would allow ISB/ISR (Incremental Static Builds/Rendering) and SSR (Server Side Rendering). It would also allow a better i18n implementation that doesn't require the two-letter folder approach and way less file generation at build time.

Is there an issue where this is discussed?

**Comment by ovflowd** [comment]
> Is there an issue where this is discussed?

Not really, the discussion is happing internally within the Foundation. But I guess you can ask about it on the cpc channel on Slack 👀

**Comment by mcollina** [comment]
Please open a public issue to discuss that. There are quite a few consideration at play regarding how releases are managed  and downloaded. @nodejs/build and @nodejs/release, and ideally @nodejs/tsc should approve. __It's not for the CPC to approve__.

**Comment by ovflowd** [comment]
@mcollina I'm not sure I understood you, the consideration for Vercel is only for the website, not for the downloads and releases, they will all stay in the current server.

Basically all pages that are handled by this repository would theoretically be handled on a Vercel App (of this repository). All remaining routes continue to be served as normally.

Also both the TSC, the Build WG and the Release WG are part of the discussions. Sorry if that wasn't clear.

**Comment by ovflowd** [comment]
In other words, the current NGINX configuration would be updated so that all routes that fall for the website would be proxied to the Vercel App.

**Comment by richardlau** [comment]
> Also both the TSC, the Build WG and the Release WG are part of the discussions. Sorry if that wasn't clear.

Not quite. Members of those groups have been involved in the discussions, but other members (such as @mcollina) are probably not aware. I do think using Vercel infrastructure needs further discussion (separate from this PR), at least visible to the whole TSC.

(And apologies, reviewing this PR is still on my list of things to do.)

**Comment by ovflowd** [comment]
Oh shame on me, I thought these discussions already happened. I understand @mcollina's confusion now 😅

Anyhow, yes, let's open an issue on /admin for this, shall we?

(Also this PR is not related to Vercel of any sorts, so this can be merged independently)

**Comment by mcollina** [comment]
> Anyhow, yes, let's open an issue on /admin for this, shall we?

Thanks! There are a few considerations to make on that topic.

**Review by richardlau — COMMENTED** [review]
RSLGTM

**Review by richardlau — APPROVED** [review]
RSLGTM

**Review by Trott — APPROVED** [review]
Rubber-stamp LGTM

**Pull request merged**
